CMSgov/mdct-mcr
# implement server-side validation

## Description

Well, we needed server-side validation -- all data needs to be validated before it hits the database. 

Changes:
- Added server-side validation in general via the `validateData` utility in _/utils/validation/validation_.
- Added server-side validation to `reportData` write call. This required:
    - Added `fetchReport` call to get the corresponding `formTemplateId`. _(/handlers/reportData/write.ts)_
- Added server-side validation (with custom server-side validation schema) to the following `write` calls:
    - Write `report` metadata on report creation or edit via `AddEditProgramModal` _(/handlers/reports/write.ts)_
    - Write `banner` data from `AddAdminBanner` _(/handlers/banners/write.ts)_
- Fixed related tests and added some more

### How to test

1. Test the full lifecycle of creating a new report, entering it, filling out and saving information. See that everything works as expected.
2. Try to break the validation by sending invalid data. (Note: The easiest way to do this is to subtly alter the server-side validation schemas, say from `string` to `number` and submit the same data. That way you don't have to disable the client-side inline field validation.)

### Changed Dependencies

Added `yup` to `package.json` in app-api. You will have to run `yarn` in there to get it up and running the first time.

## Code author checklist

## Reviewer checklist (two different people)
